File: eq_transform.cpp
#if 1
#include <opencv2\opencv.hpp>
#include <string>


using namespace cv;
using namespace std;

int sign(double x)
{ 
	if (x == 0)
	{
		return 0;
	}
	else
	{
		if (x > 0) return 1; 
		else return -1;
	}
}

int main(int argc, char* argv[])
//int eq_transform(int argc, char* argv[])
{
	double pi = 3.14159265358979323846;


	string filename = "1.jpeg";
	double ar, ap, ay;
	ar = 0/4;
	ap = -pi/4;
	ay = 0/4;
	cv::Mat img_origin = cv::imread(filename);
	cv::Size img_size = img_origin.size();
	cv::Mat img_undist = cv::Mat::zeros(img_size, CV_8UC3);
	int pnwidth = img_size.width;
	int pnheight = img_size.height;
	
	double b = 2 * pi;
	double mpdistance = pnwidth / b;
	
	double rot[2] = { mpdistance*pi, -ay*mpdistance };
	
	double m[3][3] = { cos(ar), sin(ar), 0,
		-cos(ap)*sin(ar), cos(ap)*cos(ar), sin(ap),
		sin(ar)*sin(ap), -sin(ap)*cos(ar), cos(ap)
	};

	
	{


		for (int x = 1; x <= pnheight; ++x)
		{

			for (int y = 1; y <= pnwidth; ++y)
			{
		
				double src[2] = { y , x };
				src[0] += rot[1];
			 		
				double phi = src[0] / mpdistance;
				double theta1 = src[1] / mpdistance - pi / 2;
												
				//double v[3] = { cos(phi)*cos(theta1), sin(phi)*cos(theta1), sin(theta1) };
				double v[3] = { cos(phi)*cos(theta1), sin(theta1), sin(phi)*cos(theta1), };
				double u[3] = { v[0] * m[0][0] + v[1] * m[1][0] + v[2] * m[2][0], v[0] * m[0][1] + v[1] * m[1][1] + v[2] * m[2][1], v[0] * m[0][2] + v[1] * m[1][2] + v[2] * m[2][2] };
				
				phi = atan2(u[2], u[0]);
				theta1 = asin(u[1]) + pi / 2;

				src[0] = mpdistance*(phi + sign(abs(u[0]))*pi);
				src[0] = (int)(src[0] - 0.5 + pnwidth / 2) % pnwidth;
				src[1] = (int)(mpdistance*(theta1) - 0.5 );
		
				if (src[1] < 1)
					src[1] = 1;
				if (src[1] > pnwidth / 2 )
					src[1] = pnwidth / 2 ;
				if (src[0] < 1)
					src[0] = 1;
				if (src[0] > pnwidth )
					src[0] = pnwidth ;
				
				img_undist.at<cv::Vec3b>(x-1, y-1) = img_origin.at<cv::Vec3b>(src[1], src[0]);
			}
		}
	}
	cv::imwrite("rota.jpg", img_undist);
	std::system("pause");
	return 0;
}
#endif

File: eq_transform.m
% parameters : filename, roll, yaw, pitch
% note that r,y,p in radius
% example call: 
% gyro_stabilization('image_DST.tif', 0, pi/4, pi/2);

% function gyro_stabilization(filename, r, y, p)
  filename = '2.jpeg';
  r = pi/4;
  y = 0;
  p = pi/4;
  img_origin=imread(filename);
  img_size = size( img_origin );
  img_undist = zeros( img_size );
  img_undist = uint8( img_undist );
  pnwidth = img_size(2);
  pnheight = img_size(1);
  b = 2*pi;
  mpdistance = pnwidth/b;

  rot = [(mpdistance*pi) (-y*mpdistance)];

  % rotation matrix
  m =[cos(r) sin(r) 0; -cos(p)*sin(r) cos(p)*cos(r) sin(p);sin(r)*sin(p)  -sin(p)*cos(r)  cos(p)];

%  for l0 = 1:img_size(3)
    for l1 = 1:img_size(1)
      x = l1;
      for l2 = 1:img_size(2)
        y =  l2;
        src = [y x];
%         src = src - [pnwidth/2  pnheight/2];%×ø±ê±ä»»
        %src(1) = src(1) + rot(2);
%         src(1)=sign(abs(src(1)-pnwidth/2))*((sign(src(1)-pnwidth/2)+1)*(src(1)-pnwidth/2)-sign(src(1)-pnwidth/2)*src(1))+(1-sign(abs(src(1)-pnwidth/2)))*src(1);
%         src(1)=sign(abs(src(1)+pnwidth/2))*((1-sign(src(1)+pnwidth/2))*(src(1)+pnwidth/2)+sign(src(1)+pnwidth/2)*src(1))+(1-sign(abs(src(1)+pnwidth/2)))*src(1);
%         
        phi = src(1) / mpdistance;
        theta1 = src(2)/ mpdistance - pi/2;

       
        
        v = [cos(phi)*cos(theta1), sin(phi)*cos(theta1),sin(theta1)];
        u=[(v(1)* m(1,1)+v(2)*m(2,1)+v(3)*m(3,1)) (v(1) * m(1,2) + v(2) * m(2,2) + v(3) * m(3,2)) (v(1) * m(1,3) + v(2) * m(2,3) + v(3) * m(3,3))];
        v = u;
        phi = atan2(v(2), v(1));
        theta1 = asin(v(3)) + pi/2;
        src(1) = mpdistance *(atan2(v(2), v(1)) + sign(abs(v(1)))*pi);
        src(2)= mpdistance * (asin(v(3)) + pi/2);
        %resize(2.6666666666666665000, 2.6666666666666665000)
        src(1)  = mod(src(1) + 1500, 3000);
        src = ceil(src+ [0 0]);
        
        if src(2) < 1
          src(2) =1 ;
        end
        if src(2) > pnwidth/2
          src(2) =pnwidth/2 ;
        end
        if src(1) < 1
          src(1) =1 ;
        end
        if src(1) > pnwidth
          src(1) =pnwidth ;
        end
        img_undist(x,y,:) = img_origin(src(2),src(1),:);
        %     if src(2) < 1 || src(2) > 120  || src(1) < 1  src(1) > 240
        %         img_undist(x,y,l0) = img_origin(0,0,l0);
        %     else
        %         img_undist(x,y,l0) = img_origin(src(2),src(1),l0);
        %     end
      end
    end
  
  % compare the two images in one figure
  figure(1);
  subplot(121);
  imshow(img_origin);
  subplot(122);
  imshow(img_undist);

  % save the rotated image
  [pathstr,name,ext] = fileparts(filename)
  imwrite(img_undist, [name, '_rotated', ext]);
  
% end 